R/print.data.table.R

```diff
@@ -164,3 +186,47 @@ shouldPrint = function(x) {
 #   as opposed to printing a blank line, for excluding col.names per PR #1483
 cut_top = function(x) cat(capture.output(x)[-1L], sep = '\n')
 
+# to calculate widths of data.table for PR #4074
+nchar_width = function(x) {
+  each_width = nchar(x, type = "width")
+  widths = vector("numeric", ncol(x))
+  for (i in 1L:ncol(x)){
+    widths[i] = max(each_width[, i])
+  }
+  widths
+}
+# gets the width of the data.table at each column
+#   and compares it to the console width
+dt_width = function(x, class, row.names, col.names) {
+  widths = nchar_width(x)
+  if (class) widths = ifelse(widths < 6L, 6L, widths)
+  if (col.names != "none") names = sapply(colnames(x), nchar, type = "width") else names = 0L
+  dt_widths = ifelse(widths > names, widths, names)
+  rownum_width = if (row.names) max(nchar(rownames(x), type = "width")) else 0L
```

Comment:
This one is also an expensive call we can avoid... if `x` has e.g. 1e9 rows, we would convert 1e9 integers to character, then run `nchar`. Whereas from `data.table(1:1e9)` we can see it's always printed as all digits (no scientific notation). So `if(row.names)`, `rownum_width` is `as.integer(ceiling(log10(nrow(x)))+1)`

Reply:
Could we grab the last row name assuming it will be the largest? e.g. `rownames(x)[length(x)]`? We would only need that one to add the widths.

Reply:
Sorry, not `length(x)` but `nrow(x)`? Is that too expensive of a call on tons of rows?

Reply:
Wow. Ignore this question. Just ran some benchmarks with your code.

```
as.integer(ceiling(log10(nrow(DT)))+1)
#> 4
nchar(rownames(DT)[nrow(DT)], type = "width")
#> 3
bench::system_time(as.integer(ceiling(log10(nrow(DT)))+1))
#> process 
#>    19µs 
#>    real 
#>  28.9µs 
bench::system_time(nchar(rownames(DT)[nrow(DT)], type = "width"))
#> process 
#>   203µs 
#>    real 
#>   204µs 
```

I'll put your chunk in there.

Reply:
It's even slightly faster than the simplified version using `nchar()`:

```
nchar(nrow(DT), type = "width")+1
#> 4
bench::system_time(nchar(nrow(DT), type = "width"))
#> process 
#>    29µs 
#>    real 
#>  29.4µs 
```

Reply:
`data.table` only keeps row names for consistency to `data.frame`; they're always `1:nrow(x)`, so `rownames(x)[length(x)]` is always `nrow(x)`. `nrow` will run almost instantaneously since it's an attribute of the column vectors, rather than needing to iterate over the vector to count rows, we just read the attribute. In-memory data is great!

`length(x)` is the same as `ncol(x)` -- this comes from the `list` inheritance. `data.table(a = 1, b = 2)` is "decorated" `list(a=1, b=2)`, the length of which is `2L`.

Reply:
Still can't believe how quick your code was. Would have never thought of it.